graphix/gflow.py

```diff
@@ -12,228 +12,123 @@
-    of l_k from 1), and for each measurements of qubit i we must perform corrections on
-    qubits in g(i), depending on the measurement outcome.
+    AnyMeasurement = Union[Measurement, PauliMeasurement]
 
-    For more details of gflow, see Browne et al., NJP 9, 250 (2007).
-    We use the extended gflow finding algorithm in Backens et al., Quantum 5, 421 (2021).
 
-    Parameters
-    ----------
-    graph: :class:`networkx.Graph`
-        Graph (incl. input and output)
-    iset: set
-        set of node labels for input
-    oset: set
-        set of node labels for output
-    meas_planes: dict
-        measurement planes for each qubits. meas_planes[i] is the measurement plane for qubit i.
-    mode: str
-        The gflow finding algorithm can yield multiple equivalent solutions. So there are three options
+@functools.singledispatch
+def _pp_convert(p: AnyMeasurement) -> PPlane_:  # noqa: ARG001
+    raise TypeError
 
-        - "single": Returrns a single solution
 
-        - "all": Returns all possible solutions
+@_pp_convert.register
+def _(p: Measurement) -> PPlane_:
+    if p.plane == Plane.XY:
+        return PPlane_.XY
+    if p.plane == Plane.YZ:
+        return PPlane_.YZ
+    if p.plane == Plane.XZ:
+        return PPlane_.XZ
+    typing_extensions.assert_never(p.plane)
 
-        - "abstract": Returns an abstract solution. Uncertainty is represented with sympy.Symbol objects,
-          requiring user substitution to get a concrete answer.
 
-        Optional. Default is "single".
+@_pp_convert.register
+def _(p: PauliMeasurement) -> PPlane_:
+    if p.axis == Axis.X:
+        return PPlane_.X
+    if p.axis == Axis.Y:
+        return PPlane_.Y
+    if p.axis == Axis.Z:
+        return PPlane_.Z
+    typing_extensions.assert_never(p.axis)
 
-    Returns
-    -------
-    g: dict
-        gflow function. g[i] is the set of qubits to be corrected for the measurement of qubit i.
-    l_k: dict
-        layers obtained by gflow algorithm. l_k[d] is a node set of depth d.
-    """
-    check_meas_planes(meas_planes)
-    l_k = {}
-    g = {}
-    for node in graph.nodes:
-        l_k[node] = 0
-    return gflowaux(graph, iset, oset, meas_planes, 1, l_k, g, mode=mode)
 
+def _p_convert(p: Measurement) -> Plane_:
+    if p.plane == Plane.XY:
+        return Plane_.XY
+    if p.plane == Plane.YZ:
+        return Plane_.YZ
+    if p.plane == Plane.XZ:
+        return Plane_.XZ
+    typing_extensions.assert_never(p.plane)
 
-def gflowaux(
-    graph: nx.Graph,
-    iset: set[int],
-    oset: set[int],
-    meas_planes: dict[int, Plane],
-    k: int,
-    l_k: dict[int, int],
-    g: dict[int, set[int]],
-    mode: str = "single",
-):
-    """Find one layer of the gflow.
 
-    Ref: Backens et al., Quantum 5, 421 (2021).
+def _default_construct(keys: Iterable[int]) -> dict[int, Measurement]:
+    # Random angle for safety
+    return dict.fromkeys(
+        keys,
+        Measurement(Plane.XY, 0.5014943209046647),
+    )
```

Comment:
Having a dummy value for measurement angles looks a bit awkward and is a sign that `Measurement` is not the right data structure to use in this module. I hope that we can eliminate it with the refactoring I propose in this review: use `Measurement` in `OpenGraph` and `generate_from_graph`, but use `Plane` and `PPlane` in `gflow`.